Check access to tmp folder

ClementKunz opened the pull request:
On app launch, check if the app have access to the temp folder. If it is not the case, do not start the syncs and show a didecated error message

ClementKunz committed:
fix: check access to tmp folder on app launch

ClementKunz committed:
fix: change error message

ClementKunz committed:
test: add unit test

ClementKunz committed:
fix: handle access denied of tmp directory during a sync

ClementKunz committed:
fix: add a random suffix

ClementKunz committed:
chore: update translations

ChristopheLarchier commented on src/libcommonserver/utility/utility.cpp:
```diff
@@ -667,4 +668,63 @@ bool Utility::isError500(const Poco::Net::HTTPResponse::HTTPStatus httpErrorCode
+IoError Utility::tryCreateTmpDir(const SyncName &name /*= "testDir"*/) {
+#if defined(KD_MACOS)
+    // Add random suffix to make sure the file does not already exist.
+    const auto filename = name + SyncName(CommonUtility::generateRandomStringAlphaNum());
```
This can be a problem if the maximum name size is reached.
Wouldn't it be better to test the return code of create_directory?

ClementKunz replied:
Done in bf4d53879f0f19133a05b22e8d39e0b6652dc700.
Return code of `create_directory` is checked below

luc-guyot-infomaniak commented on src/libsyncengine/update_detection/file_system_observer/remotefilesystemobserverworker.cpp:
```diff
@@ -341,7 +341,10 @@ ExitInfo RemoteFileSystemObserverWorker::getItemsInDir(const NodeId &dirId, cons
 
         // Check unsupported characters
-        if (hasUnsupportedCharacters(item.name(), item.id(), item.type())) {
+        if (const auto ioError = checkForUnsupportedCharacters(item.name(), item.id(), item.type()); !ioError) {
```
```suggestion
        if (const auto exitInfo = checkForUnsupportedCharacters(item.name(), item.id(), item.type()); !ioError) {
```

ClementKunz replied:
Done in 53ad14b019d7776aec6ae5777c3cbba673277de9

luc-guyot-infomaniak commented on src/libsyncengine/update_detection/file_system_observer/remotefilesystemobserverworker.cpp:
```diff
@@ -480,8 +483,12 @@ ExitInfo RemoteFileSystemObserverWorker::processActions(Poco::JSON::Array::Ptr a
         // Check unsupported characters
-        if (hasUnsupportedCharacters(actionInfo.snapshotItem.name(), actionInfo.snapshotItem.id(),
-                                     actionInfo.snapshotItem.type())) {
+        if (const auto ioError = checkForUnsupportedCharacters(actionInfo.snapshotItem.name(), actionInfo.snapshotItem.id(),
```
```suggestion
        if (const auto exitInfo = checkForUnsupportedCharacters(actionInfo.snapshotItem.name(), actionInfo.snapshotItem.id(),
```

ClementKunz replied:
Done in 53ad14b019d7776aec6ae5777c3cbba673277de9

ChristopheLarchier commented on src/libcommonserver/utility/utility.cpp:
```diff
@@ -667,4 +668,63 @@ bool Utility::isError500(const Poco::Net::HTTPResponse::HTTPStatus httpErrorCode
+        return ioError;
+    }
+
+    const auto filename = name + SyncName(CommonUtility::generateRandomStringAlphaNum());
```
This can be a problem if the maximum name size is reached.

ClementKunz replied:
Done in bf4d53879f0f19133a05b22e8d39e0b6652dc700
I just worried that a file with same name might remain in the tmp directory leading to a wrong error "UNICODE ...". What do you think? The file creation using `std::ofstream` does provide any feedback.

ChristopheLarchier replied:
It is quickly removed but...
At worst, if 2 jobs try to create the same file at the same time, one will error out and the file will be excluded from the synchronization. It will be synced next time.

ClementKunz replied:
Yes, not a big deal indeed 👍

luc-guyot-infomaniak commented on test/libcommonserver/utility/testutility.cpp:
```diff
@@ -537,4 +537,16 @@ void TestUtility::testUserName() {
 #endif
 }
 
+void TestUtility::testTryCreateTmpDir() {
```
It would be nice to also have failure scenarios that revolve around missing permissions (the initial problem).

ClementKunz replied:
Done in adb3d701263ce0326992c120113d4bd805958673

ClementKunz committed:
fix: fix Windows build

ChristopheLarchier commented on src/gui/parametersdialog.cpp:
You could move those 4 cases after SystemError.

ClementKunz replied:
Done in 0920db299f332bab7d374e48b76caa8895eb7de5

ClementKunz committed:
chore: rename variables

ClementKunz committed:
chore: Merge branch 'Check-access-to-tmp-folder' of github.com:Infomaniak/desktop-kDrive into Check-access-to-tmp-folder

ClementKunz committed:
chore: move case statements

ClementKunz committed:
chore: do not add random suffix to tmp file

ClementKunz committed:
chore: fix Linux test

ClementKunz committed:
test: test access denied

ClementKunz committed:
chore: Merge branch 'release/release-3.7.7' into Check-access-to-tmp-folder

# Conflicts:
#	translations/client_de.ts
#	translations/client_es.ts
#	translations/client_fr.ts
#	translations/client_it.ts

ClementKunz committed:
chore: update translations

ClementKunz committed:
test: fix test for Linux

ClementKunz committed:
test: remove unreliable test

ClementKunz committed:
chore: Merge branch 'release/release-3.7.7' into Check-access-to-tmp-folder

# Conflicts:
#	src/server/appserver.cpp

ClementKunz committed:
test: fix tests

ClementKunz committed:
test: fix test
